### PR TITLE
refactor(statistic): refactor statistic and countdown to typescript

### PR DESCRIPTION
**Pull request by ununian**
First of all, thank you for your contribution! 😄

New feature please send pull request to feature branch, and rest to master branch. Pull request will be merged after one of collaborators approve. Please makes sure that these form are filled before submitting your pull request, thank you!

[[中文版模板 / Chinese template](https://github.com/vueComponent/ant-design-vue/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / document update
- [ ] Component style update
- [ ] TypeScript definition update
- [x] Refactoring
- [ ] Code style optimization
- [ ] Branch merge
- [ ] Other (about what?)

### What's the background?

> 1. Describe the source of requirement.
> 2. Resolve what problem.
> 3. Related issue link.

### API Realization (Optional if not new feature)

> 1. Basic thought of solution and other optional proposal.
> 2. List final API realization and usage sample.
> 3. GIF or snapshot should be provided if includes UI/interactive modification.

### What's the effect? (Optional if not new feature)

> 1. Does this PR affect user? Which part will be affected?
> 2. What will say in changelog?
> 3. Does this PR contains potential break change or other risk?

### Changelog description (Optional if not new feature)

> 1. English description
> 2. Chinese description (optional)

### Self Check before Merge

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

### Additional Plan? (Optional if not new feature)

> If this PR related with other PR or following info. You can type here.

### CODE REVIEW

File: components/statistic/Countdown.tsx [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  name: 'AStatisticCountdown',
+  props: {
+    ...StatisticProps,
+    format: { type: String, default: 'HH:mm:ss' },
+  },
+  emits: ['finish'],
+  setup(props, { emit }) {
+    const { format } = props;
+
+    let countdownId: number | undefined = undefined;
```
Comment (Amour1688) [inline components/statistic/Countdown.tsx:23]: `countdownId` 不是响应式的

File: components/statistic/Countdown.tsx [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+const StatisticCountdown = defineComponent({
+  name: 'AStatisticCountdown',
+  props: {
+    ...StatisticProps,
+    format: { type: String, default: 'HH:mm:ss' },
+  },
+  emits: ['finish'],
+  setup(props, { emit }) {
+    const { format } = props;
```
Comment (Amour1688) [inline components/statistic/Countdown.tsx:21]: 在这里不能解构

File: components/statistic/Countdown.tsx (hunk 1 above)
Comment (ununian) [reply components/statistic/Countdown.tsx:23]: 这个不需要是响应式的，不会用到模板里面，只是控制定时器的

File: components/statistic/Statistic.tsx [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  props: StatisticProps,
+  setup(props, { slots }) {
+    const configProvider = inject('configProvider', defaultConfigProvider);
+    const { prefixCls: customizePrefixCls, value, valueStyle } = props;
+    const getPrefixCls = configProvider.getPrefixCls;
+    const prefixCls = getPrefixCls('statistic', customizePrefixCls);
+
+    const title = getComponentFromSetup(props, slots, 'title') as VNodeTypes;
+    const prefix = getComponentFromSetup(props, slots, 'prefix') as VNodeTypes;
+    const suffix = getComponentFromSetup(props, slots, 'suffix') as VNodeTypes;
```
Comment (Amour1688) [inline components/statistic/Statistic.tsx:33]: 这些变量的值不会随着属性不变化

Comment (ununian) [reply components/statistic/Statistic.tsx:33]: 这个应该用计算属性么？

File: components/statistic/Countdown.tsx [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+      syncTimer();
+    });
+
+    onBeforeUnmount(() => {
+      stopTimer();
+    });
+
+    return () => (
+      <Statistic
+        key={renderKey.value}
```
Comment (tangjinzhou) [inline components/statistic/Countdown.tsx:66]: 不要用key，会导致子元素全部重新 render
如果要使用composition api  试试  data-key={renderKey.value}

File: components/statistic/Statistic.tsx [hunk 5]
```diff
@@ -0,0 +1,66 @@
+import { CSSProperties, defineComponent, inject, PropType, VNodeTypes, computed } from 'vue';
+import { defaultConfigProvider } from '../config-provider';
+import { getComponentFromSetup } from '../_util/props-util';
+import StatisticNumber from './Number';
+
+export const StatisticProps = {
+  decimalSeparator: { type: String, default: '.' },
```
Comment (tangjinzhou) [inline components/statistic/Statistic.tsx:7]: 缺少 format、onFinish

File: components/statistic/Statistic.tsx [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  formatter: {
+    type: Function as PropType<(params: { value: string | number | object }) => VNodeTypes>,
+  },
+  groupSeparator: { type: String, default: ',' },
+  precision: { type: Number },
+  prefix: { type: String },
```
Comment (tangjinzhou) [inline components/statistic/Statistic.tsx:13]: prefix: PropTypes.VNodeChild

**Review by tangjinzhou — CHANGES_REQUESTED** [review]

File: components/statistic/Statistic.tsx [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+  suffix: { type: String },
+  title: { type: String },
```
Comment (Amour1688) [inline components/statistic/Statistic.tsx:15]: `props` 的声明尽量参考一下已经改完的组件吧，不要单独搞一套风格

Comment (ununian) [reply components/statistic/Statistic.tsx:15]: 嗯 ，我写的时候还没几个合并了的pr，所以不知道怎么写props

File: components/statistic/Statistic.tsx (hunk 5 above)
Comment (ununian) [reply components/statistic/Statistic.tsx:7]: 这2个不是 CountDown 才有的吗？ 我就没写在 Statistic 里面了

File: components/statistic/Number.tsx [hunk 8]
```diff
@@ -0,0 +1,52 @@
+import padEnd from 'lodash-es/padEnd';
+import { VNodeTypes } from 'vue';
+import { StatisticPropsType } from './Statistic';
+
+const StatisticNumber = (
```
Comment (Amour1688) [inline components/statistic/Number.tsx:5]: 这里可以用 `FunctionalComponent`

Comment (Amour1688) [reply components/statistic/Number.tsx:5]: 函数式组件还是需要声明 `props` 的

Comment (ununian) [reply components/statistic/Number.tsx:5]: 这里我测试了一下，如果不设置函数式组件的props，那么会把所有的attrs塞入 props，相当于 props 和 attrs 是等价的。
这里这样写 tsx 能正确识别所需的 props。如果用 attrs 的话，tsx 就不能正确识别。
和 react的高阶函数式组件很像。虽然我没在vue3的文档里面找到相关的描述。
我写的时候是参考了 RenderEmpty 的代码。

Comment (Amour1688) [reply components/statistic/Number.tsx:5]: 不设置 props 在 `sfc` 会有问题

**Comment by Amour1688** [comment]
冲突解决一下吧

**Comment by ununian** [comment]
我先整理一下

**Pull request closed without merging**